properties/base/union.py

```diff
@@ -178,19 +188,22 @@ def assert_valid(self, instance, value=None):
             value = instance._get(self.name)
             if value is None:
                 return True
+        error_messages = []
         for prop in self.props:
             try:
                 return prop.assert_valid(instance, value)
-            except GENERIC_ERRORS:
-                continue
-        message = (
-            'The "{name}" property of a {cls} instance has not been set '
-            'correctly'.format(
-                name=self.name,
-                cls=instance.__class__.__name__
-            )
-        )
-        raise utils.ValidationError(message, 'invalid', self.name, instance)
+            except GENERIC_ERRORS as err:
+                if hasattr(err, 'error_tuples'):
+                    error_messages += [
+                        err_tup.message for err_tup in err.error_tuples
+                    ]
+        if error_messages:
+            extra = 'Possible explanation:'
+            for message in error_messages:
+                extra += '\n    - {}'.format(message)
+        else:
+            extra = ''
+        self.error(instance, value, extra=extra)
```

Comment:
Worth combining this with the above? Seems like `.assert_valid` and `.validate` are sharing a lot of the same stuff. Could make it an underscore-prepended `@staticmethod` that's shared?

Reply:
Yeah, this was a big copy/paste. I'll pull it out.